File: internal/usage/pricing.go
package usage

import "strings"

type ModelPricing struct {
	InputPerMillion  float64
	OutputPerMillion float64
}

var pricingTable = map[string]ModelPricing{
	"gemini-2.5-pro":       {InputPerMillion: 1.25, OutputPerMillion: 10.0},
	"gemini-2.5-flash":     {InputPerMillion: 0.15, OutputPerMillion: 0.6},
	"gemini-3-pro-preview": {InputPerMillion: 1.25, OutputPerMillion: 10.0},
	"gemini-3-pro":         {InputPerMillion: 1.25, OutputPerMillion: 10.0},
	"gemini-3-flash":       {InputPerMillion: 0.15, OutputPerMillion: 0.6},
	"gemini-2.0-flash":     {InputPerMillion: 0.10, OutputPerMillion: 0.4},
	"claude-opus-4":        {InputPerMillion: 15.0, OutputPerMillion: 75.0},
	"claude-sonnet-4":      {InputPerMillion: 3.0, OutputPerMillion: 15.0},
	"claude-haiku-3":       {InputPerMillion: 0.25, OutputPerMillion: 1.25},
	"gpt-5":                {InputPerMillion: 2.0, OutputPerMillion: 8.0},
	"gpt-5.1":              {InputPerMillion: 2.0, OutputPerMillion: 8.0},
	"gpt-5.3":              {InputPerMillion: 2.0, OutputPerMillion: 8.0},
	"gpt-4.1":              {InputPerMillion: 2.0, OutputPerMillion: 8.0},
	"gpt-4o":               {InputPerMillion: 2.5, OutputPerMillion: 10.0},
	"o3":                   {InputPerMillion: 2.0, OutputPerMillion: 8.0},
	"o4-mini":              {InputPerMillion: 1.1, OutputPerMillion: 4.4},
}

func LookupPricing(model string) (ModelPricing, bool) {
	model = strings.TrimSpace(model)
	if model == "" || strings.EqualFold(model, "default") {
		return ModelPricing{}, false
	}
	if pricing, ok := pricingTable[model]; ok {
		return pricing, true
	}
	bestLen := 0
	var best ModelPricing
	var found bool
	for prefix, pricing := range pricingTable {
		if len(prefix) > bestLen && strings.HasPrefix(model, prefix) {
			bestLen = len(prefix)
			best = pricing
			found = true
		}
	}
	return best, found
}

func CalculateCost(inputTokens, outputTokens int, model string) float64 {
	pricing, found := LookupPricing(model)
	if !found {
		return 0
	}
	inputCost := float64(inputTokens) * pricing.InputPerMillion / 1_000_000
	outputCost := float64(outputTokens) * pricing.OutputPerMillion / 1_000_000
	return inputCost + outputCost
}

File: web/src/utils/pricing.ts
/**
 * Model pricing table — mirrors internal/tui/settings_usage.go pricingTable.
 */

interface ModelPricing {
  inputPerMillion: number;
  outputPerMillion: number;
}

const PRICING_TABLE: Record<string, ModelPricing> = {
  "gemini-2.5-pro": { inputPerMillion: 1.25, outputPerMillion: 10.0 },
  "gemini-2.5-flash": { inputPerMillion: 0.15, outputPerMillion: 0.6 },
  "gemini-3-pro-preview": { inputPerMillion: 1.25, outputPerMillion: 10.0 },
  "gemini-3-pro": { inputPerMillion: 1.25, outputPerMillion: 10.0 },
  "gemini-3-flash": { inputPerMillion: 0.15, outputPerMillion: 0.6 },
  "gemini-2.0-flash": { inputPerMillion: 0.1, outputPerMillion: 0.4 },
  "claude-opus-4": { inputPerMillion: 15.0, outputPerMillion: 75.0 },
  "claude-sonnet-4": { inputPerMillion: 3.0, outputPerMillion: 15.0 },
  "claude-haiku-3": { inputPerMillion: 0.25, outputPerMillion: 1.25 },
  "gpt-5": { inputPerMillion: 2.0, outputPerMillion: 8.0 },
  "gpt-5.1": { inputPerMillion: 2.0, outputPerMillion: 8.0 },
  "gpt-5.3": { inputPerMillion: 2.0, outputPerMillion: 8.0 },
  "gpt-4.1": { inputPerMillion: 2.0, outputPerMillion: 8.0 },
  "gpt-4o": { inputPerMillion: 2.5, outputPerMillion: 10.0 },
  "o3": { inputPerMillion: 2.0, outputPerMillion: 8.0 },
  "o4-mini": { inputPerMillion: 1.1, outputPerMillion: 4.4 },
};

/** Longest-prefix lookup, matching Go's lookupModelContextWindow pattern */
function lookupPricing(model: string): ModelPricing | null {
  if (!model) return null;
  // Exact match first
  if (PRICING_TABLE[model]) return PRICING_TABLE[model];
  // Longest prefix
  let bestLen = 0;
  let bestPricing: ModelPricing | null = null;
  for (const [prefix, pricing] of Object.entries(PRICING_TABLE)) {
    if (prefix.length > bestLen && model.startsWith(prefix)) {
      bestLen = prefix.length;
      bestPricing = pricing;
    }
  }
  return bestPricing;
}

/** Calculate cost in USD for a given usage and model */
export function calculateCost(
  inputTokens: number,
  outputTokens: number,
  model: string,
): number {
  const pricing = lookupPricing(model);
  if (!pricing) return 0;
  const inputCost = (inputTokens * pricing.inputPerMillion) / 1_000_000;
  const outputCost = (outputTokens * pricing.outputPerMillion) / 1_000_000;
  return inputCost + outputCost;
}
